File: scripts/extract-pdfs-from-empty-texts.py
#!/usr/bin/env python3

import os
import re
import shutil
from pathlib import Path

# Directories
base_dir = Path(__file__).parent.parent
texts_dir = base_dir / 'texts'
assets_dir = base_dir / 'assets' / 'pdf'
output_dir = base_dir / 'pdf-extract'

# Create output directory
output_dir.mkdir(exist_ok=True)

found_count = 0
copied_count = 0
not_found = []

print(f"Scanning markdown files in {texts_dir}...")

for md_file in texts_dir.glob('*.md'):
    content = md_file.read_text(encoding='utf-8')
    
    # Extract YAML front matter
    yaml_match = re.match(r'^---\s*\n(.*?)\n---\s*\n?(.*)$', content, re.DOTALL)
    
    if not yaml_match:
        continue
    
    front_matter, body_content = yaml_match.groups()
    
    # Check if body is empty (only whitespace)
    if body_content.strip():
        continue
    
    # Look for pdf field in YAML
    pdf_match = re.search(r'^pdf:\s*["\']?(.+?)["\']?\s*$', front_matter, re.MULTILINE)
    
    if not pdf_match:
        continue
    
    pdf_path = pdf_match.group(1).strip()
    found_count += 1
    
    print(f"\n📄 {md_file.name}")
    print(f"   PDF field: {pdf_path}")
    
    # Extract filename from path
    if pdf_path.startswith('/assets/pdf/'):
        pdf_filename = pdf_path.replace('/assets/pdf/', '')
    elif pdf_path.startswith('assets/pdf/'):
        pdf_filename = pdf_path.replace('assets/pdf/', '')
    else:
        pdf_filename = Path(pdf_path).name
    
    source_pdf_path = assets_dir / pdf_filename
    
    if source_pdf_path.exists():
        dest_pdf_path = output_dir / pdf_filename
        shutil.copy2(source_pdf_path, dest_pdf_path)
        print(f"   ✅ Copied to pdf-extract/{pdf_filename}")
        copied_count += 1
    else:
        print(f"   ❌ PDF not found at: {source_pdf_path}")
        not_found.append((md_file.name, pdf_path))

print(f"\n=== Summary ===")
print(f"Empty text files with PDF field: {found_count}")
print(f"PDFs successfully copied: {copied_count}")
print(f"PDFs not found: {found_count - copied_count}")

if not_found:
    print(f"\nFiles with missing PDFs:")
    for md_name, pdf in not_found:
        print(f"  - {md_name}: {pdf}")

File: scripts/extract-pdfs-from-empty-texts.js
const fs = require('fs');
const path = require('path');

const textsDir = path.join(__dirname, '..', 'texts');
const assetsDir = path.join(__dirname, '..', 'assets', 'pdf');
const outputDir = path.join(__dirname, '..', 'pdf-extract');

// Create output directory if it doesn't exist
if (!fs.existsSync(outputDir)) {
  fs.mkdirSync(outputDir, { recursive: true });
}

const files = fs.readdirSync(textsDir).filter(f => f.endsWith('.md'));
let foundCount = 0;
let copiedCount = 0;

console.log(`Scanning ${files.length} markdown files in texts/...`);

files.forEach(file => {
  const filePath = path.join(textsDir, file);
  const content = fs.readFileSync(filePath, 'utf8');
  
  // Extract YAML front matter
  const yamlMatch = content.match(/^---\s*\n([\s\S]*?)\n---\s*\n?([\s\S]*)$/);
  
  if (!yamlMatch) return;
  
  const [, frontMatter, bodyContent] = yamlMatch;
  
  // Check if body is empty (only whitespace)
  if (bodyContent.trim().length > 0) return;
  
  // Look for pdf field in YAML
  const pdfMatch = frontMatter.match(/^pdf:\s*["']?(.+?)["']?\s*$/m);
  
  if (!pdfMatch) return;
  
  const pdfPath = pdfMatch[1].trim();
  foundCount++;
  
  console.log(`\n📄 ${file}`);
  console.log(`   PDF field: ${pdfPath}`);
  
  // Check if PDF exists
  // Handle both absolute paths like /assets/pdf/... and relative paths
  let pdfFilename;
  if (pdfPath.startsWith('/assets/pdf/')) {
    pdfFilename = pdfPath.replace('/assets/pdf/', '');
  } else if (pdfPath.startsWith('assets/pdf/')) {
    pdfFilename = pdfPath.replace('assets/pdf/', '');
  } else {
    pdfFilename = path.basename(pdfPath);
  }
  
  const sourcePdfPath = path.join(assetsDir, pdfFilename);
  
  if (fs.existsSync(sourcePdfPath)) {
    const destPdfPath = path.join(outputDir, pdfFilename);
    fs.copyFileSync(sourcePdfPath, destPdfPath);
    console.log(`   ✅ Copied to pdf-extract/${pdfFilename}`);
    copiedCount++;
  } else {
    console.log(`   ❌ PDF not found at: ${sourcePdfPath}`);
  }
});

console.log(`\n=== Summary ===`);
console.log(`Empty text files with PDF field: ${foundCount}`);
console.log(`PDFs successfully copied: ${copiedCount}`);
console.log(`PDFs not found: ${foundCount - copiedCount}`);
